ctapipe/io/eventfilereader.py

```diff
@@ -148,216 +139,74 @@ def check_file_compatibility(file_path):
-            compatible = False
-        return compatible
-
-    @property
-    def num_events(self):
-        self.log.info("Obtaining number of events in file...")
-        if self._num_events:
-            pass
-        else:
-            self._num_events = len(self.event_id_list)
-        self.log.info("Number of events inside file = {}"
-                      .format(self._num_events))
-        return self._num_events
-
-    @property
-    def event_id_list(self):
-        self.log.info("Retrieving list of event ids...")
-        if self._event_id_list:
-            pass
-        else:
-            self.log.info("Building new list of event ids...")
-            ids = hessio_get_list_event_ids(self.input_path,
-                                            max_events=self.max_events)
-            self._event_id_list = ids
-        self.log.info("List of event ids retrieved.")
-        return self._event_id_list
-
-    def read(self, allowed_tels=None, requested_event=None,
-             use_event_id=False):
+        generator
         """
-        Read the file using the appropriate method depending on the file origin
-
-        Parameters
-        ----------
-        allowed_tels : list[int]
-            select only a subset of telescope, if None, all are read. This can
-            be used for example emulate the final CTA data format, where there
-            would be 1 telescope per file (whereas in current monte-carlo,
-            they are all interleaved into one file)
-        requested_event : int
-            Seek to a paricular event index
-        use_event_id : bool
-            If True ,'requested_event' now seeks for a particular event id
-            instead of index
+        for event in self._generator():
+            if self.max_events and event.count >= self.max_events:
+                break
+            yield event
 
-        Returns
-        -------
-        source : generator
-            A generator that can be iterated over to obtain events
-        """
+    def __enter__(self):
+        return self
 
-        # Obtain relevent source
-        self.log.debug("Reading file...")
-        if self.max_events:
-            self.log.info("Max events being read = {}".format(self.max_events))
-        source = hessio_event_source(self.input_path,
-                                     max_events=self.max_events,
-                                     allowed_tels=allowed_tels,
-                                     requested_event=requested_event,
-                                     use_event_id=use_event_id)
-        self.log.debug("File reading complete")
-        return source
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
 
 
-# External Children
-try:
-    from targetpipe.io.eventfilereader import TargetioFileReader, \
-        ToyioFileReader
-except ImportError:
-    pass
+# EventFileReader imports so that EventFileReaderFactory can see them
+import ctapipe.io.hessiofilereader
```

Comment:
I don't think this should be necessary. If this is necessary, it basically forbids external implementations of  `EventFileReader`, which is not good.

Reply:
I'd like to find an alternative approach to allow the EventFileReaderFactory to find all sub-classes of EventFileReader, but this is the only option I know so far.

I believe it would require a restructuring of how EventFileReaderFactory works, which is something we should do soon, but is a topic for a new PR.